text/0000-remove-stacks-mixins.md

```diff
@@ -0,0 +1,83 @@
+
+# Summary
+[summary]: #summary
+
+*NOTE: This proposal is part of a larger initiative to reduce complexity originally outlined in https://github.com/buildpacks/rfcs/pull/167*
+
+This RFC proposes that we remove the "stack" and "mixin" concepts from the project and replace them with existing constructs in the container image ecosystem such as base images, Dockerfiles, and OS packages.
+
+# Motivation
+[motivation]: #motivation
+
+The "stack" and "mixin" concepts add unnecessary complexity to the project and make it difficult for new users and contributors to understand how buildpacks work. Compatibility guarantees that are strongly enforced by the stack contract could be replaced with metadata validations and warnings.
+
+# What it is
+[what-it-is]: #what-it-is
+
+Summary of changes:
+- Replace mixins with a [CycloneDX](https://cyclonedx.org)-formatted list of packages.
+- Replace stack metadata (including stack IDs) with canonical OS metadata.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+## Base Image Metadata
+
+Instead of a stack ID, runtime and build-time base images are labeled with the following canonicalized metadata:
+- OS (e.g., "linux", `$GOOS`)
+- Architecture (e.g., "x86_64", `$GOARCH`)
+- Distribution (optional) (e.g., "ubuntu", `$ID`)
+- Version (optional) (e.g., "18.04", `$VERSION_ID`)
+
+For Linux-based images, each field should be canonicalized against values specified in `/etc/os-release` (`$ID` and `$VERSION_ID`).
+
+The `stacks` list in `buildpack.toml` is replaced by a `platforms` list, where each entry corresponds to a different buildpack image that is exported into a [manifest index](https://github.com/opencontainers/image-spec/blob/master/image-index.md). Each entry may contain multiple valid values for Distribution and/or Version, but only a single OS and Architecture. Each entry may also contain a list of package names (as PURL URLs without versions or qualifiers) that specify detect-time and build-time (but not runtime) OS package dependencies. Buildpacks may express runtime OS package dependencies during detection (see "Runtime Base Image Requirements" below).
+
+App image builds fail if the build image and selected run image have mismatched metadata. We may consider introducing a flag to skip this validation.
+
+When an app image is rebased, `pack rebase` will fail if the new run image and previous run image have mismatched metadata. This check may be skipped for Distribution and Version by passing a new `--force` flag to `pack rebase`.
+
+## Mixins
+
+The mixins label on each base image is replaced by a layer in each base image containing a single file consisting of a [CycloneDX](https://cyclonedx.org)-formatted list of packages. Each package entry has a [PURL](https://github.com/package-url/purl-spec)-formatted ID that uniquely identifies the package.
```

Comment:
Would it be easier just to replace mixins with `purls` in this RFC and leave CycloneDx out for a separate RFC that introduces stack/base image BOMs and how they intereact with the final image?

The rest of this RFC should still be valid with this change.

So we replace stacks with the above information and mixins defined by the CNB project for official stacks with `purls`?

Reply:
I worry that would leave us supporting three possible formats instead of two. Maybe we should block this RFC on SBoM format consensus?

Reply:
I would personally like to decouple base image package requirements from the sBOM. I.E I think platforms should be able to validate againt a flat list of packages in a label or file, without having to parse CycloneDx

Reply:
Will add example minimal CycloneDX JSON for feedback

Reply:
```json

{
  "bomFormat": "CycloneDX",
  "specVersion": "1.3",
  "version": 1,
  "components": [
    {
      "type": "library",
      "name": "tomcat-catalina",
      "version": "9.0.14",
      "purl": "pkg:maven/org.apache.tomcat/tomcat-catalina@9.0.14"
    },
    {
      "type": "library", 
      "name": "card-verifier",
      "version": "1.0.2",
      "purl": "pkg:maven/org.acme/card-verifier@1.0.2?repository_url=repo.acme.org/maven"
    }
  ]
}
```

This is the minimal required set of fields. purl is optional for cdx but we need it.

Reply:
Looks quite minimal to me.

@ekcasey Given that a separate list may end up as a list of objects anyways, and given that we'd likely need to put both lists in files in a layer, do you still feel like a separate list is worth it?

Reply:
Resolved by getting rid of mixins entirely.